maven/WFLY-14934_Server_Plugin.adoc

```diff
@@ -0,0 +1,213 @@
+and bash launch scripts. At startup time, a complex CLI script generation that impacts the server startup time is mandatorily executed.
+Having this level of configurability should be a user choice and not be enforced by the WildFly s2i image.
+
+To solve these pain points, we are removing in WFLY-14936 the server installation and its configuration aspects from the image.
+
+The WildFly S2I image becomes a generic image allowing the installation and execution of any provisioned WildFly server.
+
+Due to this removal, we need a way for customers to provision the WildFly server to be installed inside the new image 
+when building their application. This WildFly server will run the built application inside the image.
+
+To do so we are defining Maven tooling that handles the Galleon provisioning of the server, the application deployment and the configuration aspects.
+
+== Requirements
+
+The existing https://github.com/wildfly/wildfly-maven-plugin/[WildFly Maven plugin] has been designed to interact (mainly) with a running server instance. 
+We are here proposing to evolve the capabilities of this plugin for one to be able to build a fully configured server containing application deployment.
+ 
+This proposal adds a new kind of installation that is operated and controlled by the pom.xml file.  The pom.xml file contains all the information needed
+to control the server and application content and versioning.
+
+=== Requirements for cloud integration
+
+* The new ``package`` goal that this proposal adds to the Maven plugin is the goal to use during s2i build. 
+Other goals (e.g.: ``provisioning``, ``execute-commands``, ...) are not expected to be used.
+* The plugin can be used in a "cloud build" context. In particular, the plugin must be usable with:
+** WildFly s2i V2 builder image for s2i in binary and source (see: https://issues.redhat.com/browse/WFLY-14936[WFLY-14934])
+** JKube s2i and docker binary builds (see https://issues.redhat.com/browse/WFLY-14938[WFLY-14938])
+** Other technologies such as Buildpacks.
+* The generated server can be installed and run successfully inside a WildFly runtime image v2 defined 
+   in https://issues.redhat.com/browse/WFLY-14936[WFLY-14936].
+
+=== Hard Requirements
+
+==== Plugin requirements
+
+* The Maven plugin produces inside the project target dir a directory (named ```server``` by default) that contains the server and any deployment.
+
+==== WildFly Maven Plugin new goals
+
+* ```wildfly:package```: Galleon provisioning of a server, CLI script execution, copy content to the server, deploy application (if present). 
+* ```wildfly:provision```: Galleon provisioning of a server.
+
+In addition to these 2 new goals, we are replacing the existing automatic download and unzip of wildfly-dist by a Galleon provisioning of 
+WildFly. This change impacts ```run``` and ```start```goals).
+ 
+==== Galleon related Plugin configuration items
```

Comment:
Why are the configuration options not same as in the bootable JAR plugin? Example:

| Bootable JAR | Provisioning              |
|--------------|---------------------------|
| log-time     | log-provisioning-time     |
| offline      | offline-provisioning      |
| record-state | record-provisioning-state |
| cli-sessions | packaging-scripts         |

Reply:
please read https://github.com/wildfly/wildfly-proposals/pull/422#discussion_r713113618 and subsequents comments

Reply:
@jmesnil So now it is "descriptive name" vs "cross-plugin compatible name"?

Reply:
Since when "cross-plugin compatible name" is a thing? It is more important that the parameters added to this maven-wildlfy-plugin are consistent with existing parameters of the same plugin.

Reply:
Thank you @jmesnil ! "Consistency" was the word I should've used. The reason why I asked this is that users who already are using bootable JAR plugin might stumble upon this plugin too. They might also use them in the same `pom.xml`. I can imagine it would be confusing for some of them to set two "similar" plugins differently...

Reply:
WDYT about introducing the "new" names in bootable JAR plugin in future @jmesnil @jfdenise?

Reply:
There is no requirement to "backport" the name of the bootable jar plugin

Reply:
@jfdenise Good point, thanks. I created a new WFWIP tracker https://issues.redhat.com/browse/WFWIP-407